clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3506,9 +3506,26 @@ class SYCLKernelNameTypeVisitor
             IsInvalid = true;
             return;
           }
-          // Check if the declaration is completely defined within a
-          // function or class/struct.
-          if (Tag->isCompleteDefinition()) {
+
+          // Diagnose used types without complete definition i.e.
+          //   int main() {
+          //     class KernelName1;
+          //     parallel_for<class KernelName1>(..);
+          //   }
+          // This case can only be diagnosed during host compilation because the
+          // integration header is required to distinguish between the invalid
+          // code (above) and the following valid code:
+          //   int main() {
+          //     parallel_for<class KernelName2>(..);
+          //   }
+          // The device compiler forward declares both KernelName1 and
+          // KernelName2 in the integration header as ::KernelName1 and
+          // ::KernelName2. The problem with the former case is the additional
```

Comment:
So note that this (the ::KernelName1 and ::KernelName2) isn't exactly true, if the function is in a namespace, we try our best to get the name in as close of a namespace as possible (though your example uses 'main', so thats not exactly a concern).